funsor/sum_product.py

```diff
@@ -93,6 +93,91 @@ def partial_sum_product(sum_op, prod_op, factors, eliminate=frozenset(), plates=
+        plate_to_step=dict()):
+    """
+    Modified partial sum-product that supports contraction of markov sites.
+
+    :param ~funsor.ops.AssociativeOp sum_op: A semiring sum operation.
+    :param ~funsor.ops.AssociativeOp prod_op: A semiring product operation.
+    :param factors: A collection of funsors.
+    :type factors: tuple or list
+    :param frozenset eliminate: A set of free variables to eliminate,
+        including both sum variables and product variable.
+    :param dict plate_to_step: A dict mapping previous variables to current variables.
+        This can contain multiple pairs of prev->curr variable names.
+    :return: a list of partially contracted Funsors.
+    :rtype: list
+    """
+    assert callable(sum_op)
+    assert callable(prod_op)
+    assert isinstance(factors, (tuple, list))
+    assert all(isinstance(f, Funsor) for f in factors)
+    assert isinstance(eliminate, frozenset)
+    assert isinstance(plate_to_step, dict)
+    plates = frozenset(plate_to_step.keys())
+    sum_vars = eliminate - plates
+    prod_vars = eliminate.intersection(plates)
+    markov_sum_vars = frozenset()
+    for step in plate_to_step.values():
+        markov_sum_vars |= frozenset(step.keys()) | frozenset(step.values())
+    markov_sum_vars &= sum_vars
+    markov_prod_vars = frozenset(k for k, v in plate_to_step.items() if v and k in eliminate)
+
+    var_to_ordinal = {}
+    ordinal_to_factors = defaultdict(list)
+    for f in factors:
+        ordinal = plates.intersection(f.inputs)
+        ordinal_to_factors[ordinal].append(f)
+        for var in sum_vars.intersection(f.inputs):
+            var_to_ordinal[var] = var_to_ordinal.get(var, ordinal) & ordinal
+
+    ordinal_to_vars = defaultdict(set)
+    for var, ordinal in var_to_ordinal.items():
+        ordinal_to_vars[ordinal].add(var)
+
+    results = []
+    while ordinal_to_factors:
+        leaf = max(ordinal_to_factors, key=len)
+        leaf_factors = ordinal_to_factors.pop(leaf)
+        leaf_reduce_vars = ordinal_to_vars[leaf]
+        for (group_factors, group_vars) in _partition(leaf_factors, leaf_reduce_vars):
+            # contract non markov vars
+            nonmarkov_vars = group_vars - markov_sum_vars
+            f = reduce(prod_op, group_factors).reduce(sum_op, nonmarkov_vars)
+            # eliminate markov vars
+            # this is the only addinional part to partial_sum_product
+            markov_vars = group_vars.intersection(markov_sum_vars)
+            if markov_vars:
+                time = markov_prod_vars.intersection(f.inputs)
+                if not len(time) == 1:
+                    raise ValueError("intractable!")
+                for v in sum_vars.intersection(f.inputs):
+                    if time.issubset(var_to_ordinal[v]) and var_to_ordinal[v] < leaf:
+                        raise ValueError("intractable!")
+                # TODO: allow sequential_sum_product accept time as frozenset
```

Comment:
Can you explain this TODO?

Reply:
This is irrelevant to this PR. I removed it.
(I think it would be simpler to pass just the name of the time dim to the `sequential_sum_product` instead of `time_var`)